CHANGELOG.md

```diff
@@ -308,8 +308,10 @@ Additions to existing modules
   take-take             : take n (take m xs) ≡ take (n ⊓ m) xs
   take-drop             : take n (drop m xs) ≡ drop m (take (m + n) xs)
   zip-unzip             : uncurry′ zip ∘ unzip ≗ id
-  unzipWith-zipWith     : length xs ≡ length ys → unzipWith f (zipWith g xs ys) ≡ (xs , ys)
+  unzipWith-zipWith     : f ∘ uncurry′ g ≗ id → length xs ≡ length ys → unzipWith f (zipWith g xs ys) ≡ (xs , ys)
   unzip-zip             : length xs ≡ length ys → unzip (zip xs ys) ≡ (xs , ys)
+  mapMaybe-++           : mapMaybe f (xs ++ ys) ≡ mapMaybe f xs ++ mapMaybe f ys
+  unzipWith-++          : unzipWith f (xs ++ ys) ≡ zip _++_ _++_ (unzipWith f xs) (unzipWith f ys)
```

Comment:
Can these be analysed using `foldr-++` etc.?

Reply:
ooooh, good point, let me see

Reply:
oh, I see. do you have context about the "because pattern-matching definitions are generative" bit? i would like to learn more about why defining operations in Data.List.Base in terms of folds is a bad idea

Reply:
What? Defining things in terms of folds should be a *good* idea!

Reply:
Apologies for not being clear... but for details about generativity, you probably need to look at [the documentation](https://agda.readthedocs.io/en/latest/language/lambda-abstraction.html), which, somewhat frustratingly, doesn't make clear the relationship with [ordinary function definitions](https://agda.readthedocs.io/en/latest/language/function-definitions.html)...

That said, I have previously suggested refactoring the library to express (almost) all the recursions in `Data.List.Base` in terms of `foldr`/`foldl`, but IIRC, that was... discouraged. 

The following simple test suggests that, indeed, one should proceed with caution:
```agda
rev : List A → List A
rev = foldl (flip _∷_) []

rev′ : List A → List A
rev′ xs = go xs [] -- inline the definition using reverseAcc
  where
  go : List A → List A → List A
  go [] ys = ys
  go (x ∷ xs) ys = go xs (x ∷ ys)

test : rev ≡ rev′
test = refl
{- fails with:
foldl (λ y x → x ∷ y) [] x != Data.List.Base.go x x [] of type
List _A_1900
when checking that the expression refl has type rev ≡ rev′
-}
```

Reply:
So I'm not saying it's a 'bad' idea, only that one changes the *intensional* semantics, even if not the *extensional* such. Sigh. 
For better or worse, the library privileges definitions made in direct pattern-matching style rather than by specialising higher-order operations...

Reply:
See also #2267 / #2269 etc. which *do* change the intensional semantics (wrt strictness), but arguably towards 'better' definitions...

Reply:
Last word: if #2355 get merged, we can fix up `mapMaybe-++` etc. downstream in #2361 assuming that gets approval...